compliance-monitor/monitor.py

```diff
@@ -421,36 +430,42 @@ async def post_report(
     sep += len(SEP)
     signature = body_text[:sep - 1]  # do away with the ampersand!
     body_text = body_text[sep:]
-    json_text = None
+    try:
+        ssh_validate(keys, signature, body_text)
+    except ValueError:
+        raise HTTPException(status_code=401)
+
+    json_texts = []
     if content_type.endswith('-yaml'):
         yaml = ruamel.yaml.YAML(typ='safe')
-        document = yaml.load(body_text)
-        json_text = json.dumps(document, cls=TimestampEncoder)
+        documents = yaml.load_all(body_text)
+        json_texts = [json.dumps(document, cls=TimestampEncoder) for document in documents]
     elif content_type.endswith("-json"):
-        document = json.loads(body_text)
-        json_text = body_text
+        documents = [json.loads(body_text)]
+        json_texts = [body_text]
     else:
         raise HTTPException(status_code=500)
-    rundata = document['run']
-    uuid, subject, checked_at = rundata['uuid'], document['subject'], document['checked_at']
-    check_role(account, subject, ROLES['append_any'])
-    with conn.cursor() as cur:
-        keys = db_get_keys(cur, subject)
-    try:
-        ssh_validate(keys, signature, body_text)
-    except ValueError:
-        raise HTTPException(status_code=401)
-    scopeuuid = document['spec']['uuid']
+
+    allowed_subjects.add(account[0])
+    for document in documents:
+        check_role(account, document['subject'], ROLES['append_any'])
+        if document['subject'] not in allowed_subjects:
+            raise HTTPException(status_code=401, detail="delegation problem?")
+
     with conn.cursor() as cur:
-        try:
-            reportid = db_insert_report(cur, uuid, checked_at, subject, json_text, content_type, body)
-        except UniqueViolation:
-            raise HTTPException(status_code=409, detail="Conflict: report already present")
-        for version, vdata in document['versions'].items():
-            for check, rdata in vdata.items():
-                result = rdata['result']
-                approval = 1 == result  # pre-approve good result
-                db_insert_result2(cur, checked_at, subject, scopeuuid, version, check, result, approval, reportid)
+        for document, json_text in zip(documents, json_texts):
+            rundata = document['run']
+            uuid, subject, checked_at = rundata['uuid'], document['subject'], document['checked_at']
+            scopeuuid = document['spec']['uuid']
+            try:
+                reportid = db_insert_report(cur, uuid, checked_at, subject, json_text, content_type, body)
+            except UniqueViolation:
+                raise HTTPException(status_code=409, detail="Conflict: report already present")
```

Comment:
What is the intended behavior with the other reports, which did not fail until the violation? Just asking if we would need some kind of all-or-nothing transactional style here instead.

Reply:
This whole thing is inside one database transaction, so we do have that already, or did I miss something?

Reply:
Ok, I think I missed it. I was confused because there is only an explicit commit at the end in the happy case.

So we have an implicit start of the transaction because of `Depends(get_conn)`. And with the `finally` in `get_conn`, the connection is closed after `post_report` returns and thus any non-committed transaction should be rolled back. Right?

Reply:
Yes, the get_conn calls conn.close at the end, and without an explicit commit, this will roll back everything.

Reply:
The implicit start (begin transaction) is being managed by psycopg, I think, with the first statement that is sent to the database: https://www.psycopg.org/psycopg3/docs/basic/transactions.html